FSAE.X/FSAE_spi.c

```diff
@@ -0,0 +1,336 @@
+  CFGCONbits.IOLOCK = 0;
+  TRISEbits.TRISE5 = INPUT;
+  SDI2Rbits.SDI2R = 0b0110; // RPE5
+  TRISCbits.TRISC1 = OUTPUT;
+  RPC1Rbits.RPC1R = 0b0110; // SDO2
+  CFGCONbits.IOLOCK = 1;
+
+  // Initialize SCK2
+  TRISGbits.TRISG6 = OUTPUT;
+
+  // Disable interrupts
+  IEC4bits.SPI2EIE = 0;
+  IEC4bits.SPI2RXIE = 0;
+  IEC4bits.SPI2TXIE = 0;
+
+  // Disable SPI2 module
+  SPI2CONbits.ON = 0;
+
+  // Clear receive buffer
+  uint32_t readVal = SPI2BUF;
+
+  // Use standard buffer mode
+  SPI2CONbits.ENHBUF = 0;
+
+  /**
+   * F_SCK = F_PBCLK2 / (2 * (SPI2BRG + 1))
+   * SPI2BRG = (F_PBCLK2/(2*F_SCK))-1
+   */
+
+  // Set the baud rate (see above equation, ceil will round to lower freq)
+  SPI2BRG = ceil((PBCLK2/(2.0*mhz))-1);
+
+  SPI2STATbits.SPIROV = 0;
+
+  SPI2CONbits.MCLKSEL = 0; // Master Clock Enable bit (PBCLK2 is used by the Baud Rate Generator)
+  SPI2CONbits.SIDL = 0;    // Stop in Idle Mode bit (Continue operation in Idle mode)
+  SPI2CONbits.MODE32 = size & 0x20;  // 32/16-Bit Communication Select bits (8-bit)
+  SPI2CONbits.MODE16 = size & 0x10;  // 32/16-Bit Communication Select bits (16-bit)
+  SPI2CONbits.DISSDI = 0; // receiver to ignore the unused bit slots
+  SPI2CONbits.DISSDO = 0; // transmit unused bit slots with logic level 0
+  SPI2CONbits.MSTEN = 1;   // Master Mode Enable bit (Master mode)
+  SPI2CONbits.CKE = 1;     // SPI Clock Edge Select (Serial output data changes on transition from active clock state to idle clock state)
+  SPI2CONbits.SMP = 0;     // SPI Data Input Sample Phase (Input data sampled at middle of output time)
+  SPI2CONbits.CKP = 1;     // Clock Polarity Select (Idle state for clock is a high level)
+
+  // Enable SPI2 module
+  SPI2CONbits.ON = 1;
+
+  lock_config();
+}
+
+void init_spi3(double mhz, int size) {
+  unlock_config();
+
+  // Initialize SDI3/SDO3 PPS pins
+  CFGCONbits.IOLOCK = 0;
+  TRISBbits.TRISB5 = INPUT;
+  SDI3Rbits.SDI3R = 0b1000; // RPB5
+  TRISBbits.TRISB3 = OUTPUT;
+  RPB3Rbits.RPB3R = 0b0111; // SDO3
+  CFGCONbits.IOLOCK = 1;
+
+  // Initialize SCK3
+  TRISBbits.TRISB14 = OUTPUT;
+
+  // Disable interrupts
+  IEC4bits.SPI3EIE = 0;
+  IEC4bits.SPI3RXIE = 0;
+  IEC4bits.SPI3TXIE = 0;
+
+  // Disable SPI3 module
+  SPI3CONbits.ON = 0;
+
+  // Clear receive buffer
+  uint32_t readVal = SPI3BUF;
+
+  // Use standard buffer mode
+  SPI3CONbits.ENHBUF = 0;
+
+  /**
+   * F_SCK = F_PBCLK2 / (2 * (SPI3BRG + 1))
+   * SPI3BRG = (F_PBCLK2/(2*F_SCK))-1
+   */
+
+  // Set the baud rate (see above equation, ceil will round to lower freq)
+  SPI3BRG = ceil((PBCLK2/(2.0*mhz))-1);
+
+  SPI3STATbits.SPIROV = 0;
+
+  SPI3CONbits.MCLKSEL = 0; // Master Clock Enable bit (PBCLK2 is used by the Baud Rate Generator)
+  SPI3CONbits.SIDL = 0;    // Stop in Idle Mode bit (Continue operation in Idle mode)
+  SPI3CONbits.MODE32 = size & 0x20;  // 32/16-Bit Communication Select bits (8-bit)
+  SPI3CONbits.MODE16 = size & 0x10;  // 32/16-Bit Communication Select bits (16-bit)
```

Comment:
Should probably remove the `(16-bit)` and `(8-bit` to keep them consistent with the other register settings.